### PR TITLE
Unit tests for drawing_classifier

### LINKED ISSUE #1443: Add Unit Tests for Drawing Classifier

### PR DESCRIPTION
**Pull request by shantanuchhabra**
Fixes #1443.

### CODE REVIEW

**Comment by shantanuchhabra** [comment]
As of right now, there is one failing unit test related to save/load. Marked it as an xfail. I'm conflicted as to whether the fix should come as part of this PR or a different PR. Open to suggestions.

This PR is open for review.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 1]
```diff
@@ -430,9 +454,15 @@ def evaluate(self, dataset, metric='auto', output_type='dict'):
         
         return ret
 
-    def predict(self, dataset):
+    def predict(self, data):
         """ 
         Docfix coming soon
         """
-        predicted = self._predict_with_probabilities(dataset)
+        if type(data) == _tc.SArray:
```
Comment (nickjong) [inline src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Should use `isinstance`?

**Comment by nickjong** [comment]
One high-level question: how much total running time does this add to our Python unit tests?

**Comment by shantanuchhabra** [comment]
@nickjong The tests take about 47 seconds to run (including the time taken for the pre-trained model tests, which are disabled right now). A lot of that time is setting up classes and checking for SFrame equality in the `evaluate` tests.

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 2]
```diff
@@ -176,19 +194,22 @@ def _get_native_state(self):
...
-        net = _Model(num_classes = len(state['classes']), prefix = 'model0_')
+        net = _Model(num_classes = len(state['classes']), prefix = 'drawing_')
         ctx = _mxnet_utils.get_mxnet_context(max_devices=state['batch_size'])
         net_params = net.collect_params()
         _mxnet_utils.load_net_params_from_state(
             net_params, state['_model'], ctx=ctx 
             )
         state['_model'] = net
+        if state['_class_type'] == "int":
+            state['classes'] = list(map(int, state['classes']))
```
[the remark this answers is not in the document]
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: We need this since when saved models are loaded, classes of type int are loaded as floats -- something that we believe is part of a deeper issue involving the `CustomModel` in `_model.py` that calls into `turicreate.connect.main`. I discussed this with @srikris last night so I'll let him chime in if I'm not articulating something properly. :-) Should I file a follow-up issue to get rid of this hack?

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py [hunk 3]
```diff
@@ -161,9 +178,10 @@ def update_progress(cur_loss, iteration):
 
 class DrawingClassifier(_CustomModel):
 
+    _PYTHON_DRAWING_CLASSIFIER_VERSION = 0
```
[the remark this answers is not in the document]
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Changed to 1.

File: src/unity/python/turicreate/test/test_drawing_classifier.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+            num_epochs = 10)
+        self.stroke_model = _tc.drawing_classifier.create(
+            self.stroke_sf, 
+            feature = self.feature, 
+            target = self.target, 
+            num_epochs = 1)
+        self.trains = [self.check_cross_sf, self.stroke_sf]
+        self.models = [self.check_cross_model, self.stroke_model]
+
+    def test_create_with_missing_feature(self):
```
[the remark this answers is not in the document]
Comment (shantanuchhabra) [reply src/unity/python/turicreate/test/test_drawing_classifier.py]: @nickjong it's true that those two models are ignored in the tests that try to create a model with faulty data. Should I add a `test_create` that creates the two models from `setUpClass` and sets them for future tests? Of course, `pytest` runs tests in alphabetical order so I will have to make sure that this test runs after the `test_create_with_*` and before the tests that need the two models. I personally think it might be okay for the two models to be prepared in `setUpClass` because they are being used in about 14 of 20 tests, but I'm open to suggestions. Let me know!

[the remark this answers is not in the document]
Comment (nickjong) [reply src/unity/python/turicreate/test/test_drawing_classifier.py]: Sorry, I think I was confused, and didn't appreciate that `setUpClass` only runs once total, and its results are shared across individual tests

**Comment by nickjong** [comment]
Do we have no way to repro the crash outside of Travis? It sucks to have to wait for each build from scratch, along with running all the prior tests. This will also make it harder to use GitHub's functionality to track which commits each reviewer has already reviewed.

If you do need to debug in Travis, could you mark this PR as a WIP and otherwise let us know when to review changes again?

File: src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shantanuchhabra) [reply src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py]: Fixed.

File: src/unity/toolkits/drawing_classifier/data_preparation.cpp [hunk 5]
```diff
@@ -432,22 +432,18 @@ flex_image convert_stroke_based_drawing_to_bitmap(
...
-        int row_number = sframe_row[1];
-        flex_list current_stroke_based_drawing = strokes.to<flex_list>();
-        flex_image current_bitmap = convert_stroke_based_drawing_to_bitmap(
-            current_stroke_based_drawing, row_number);
-        return current_bitmap;
-    }, flex_type_enum::IMAGE);
 
+    std::vector<flexible_type> bitmaps;
+    for (int row_number = 0; row_number < (int)data.size(); row_number++) {
+        const flexible_type &strokes = data[feature][row_number];
```
Comment (nickjong) [inline src/unity/toolkits/drawing_classifier/data_preparation.cpp]: Use gl_sframe range iterator to iterate through an SFrame

File: src/unity/python/turicreate/test/test_drawing_classifier.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+    Build an SFrame from 10 saved drawings.
+    '''
+    drawings_dir = _os.path.join(
+        _os.path.dirname(_os.path.realpath(__file__)), "drawings")
+    sf = _tc.image_analysis.load_images(drawings_dir, with_path=True)
+    sf = sf.rename({"image": "drawing", "path": "label"})
+    sf["label"] = sf["label"].apply(
+        lambda filepath: _os.path.splitext(
+                            _os.path.basename(filepath)
+                            )[0][:-1]
```
[the remark this answers is not in the document]
Comment (shantanuchhabra) [reply src/unity/python/turicreate/test/test_drawing_classifier.py]: The files are named 'check0', 'check1', 'check2', ..., 'cross0', 'cross1', 'cross2', ... and we want the class label to be 'check' and 'cross'. Hence, the `[:-1]`. I'll add a comment explaining it.

File: src/unity/toolkits/drawing_classifier/data_preparation.cpp (hunk 5 above)
Comment (shantanuchhabra) [reply src/unity/toolkits/drawing_classifier/data_preparation.cpp]: Done.

File: src/unity/toolkits/drawing_classifier/data_preparation.cpp [hunk 7]
```diff
@@ -429,17 +429,29 @@ flex_image convert_stroke_based_drawing_to_bitmap(
...
         flex_list current_stroke_based_drawing = strokes.to<flex_list>();
-        bitmaps.push_back(
-            convert_stroke_based_drawing_to_bitmap(
-                current_stroke_based_drawing, row_number));
+        bitmaps.push_back(convert_stroke_based_drawing_to_bitmap(
+            current_stroke_based_drawing, row_number)
+        );
+        row_number++;
     }
     gl_sarray bitmaps_sarray(bitmaps);
```
Comment (nickjong) [inline src/unity/toolkits/drawing_classifier/data_preparation.cpp]: Note that you could also use `gl_sarray_writer`, defined in `gl_sarrah.hpp` to avoid storing all the bitmaps in main memory before attempting to construct the SArray. That would also allow us to be more explicit here about the column type

**Comment by shantanuchhabra** [comment]
> What portion of your last commit, the merge from master, should be reviewed? Or what portion of it interacts with this commits that I should be reviewing?

@nickjong There is nothing in the merge from master that interacts with code here or needs to be reviewed as part of this PR. The only change that needed to be reviewed was e703ca1cb5f482a448e582eed632621dc43ad8ea
